src/sinks/gcp/stackdriver_metrics.rs

```diff
@@ -92,162 +96,325 @@ fn default_endpoint() -> String {
+            Ok(auth) => auth,
+            Err(e) => {
+                error!(message = "Fatal gcp_stackdriver_metrics error.", %e);
+                return Err(());
+            }
+        };
 
-    fn build_encoder(&self) -> Self::Encoder {
-        StackdriverMetricsEncoder
-    }
+        let uri = if let Ok(uri) = self.config.endpoint.as_str().parse::<http::Uri>() {
+            uri
+        } else {
+            error!(message = "Error when parsing uri.", %self.config.endpoint);
+            return Err(());
+        };
+
+        let channel = Channel::builder(uri)
+            .tls_config(tonic::transport::ClientTlsConfig::new())
+            .map_err(|e| {
+                error!(message = "Fatal gcp_stackdriver_metrics error.", %e);
+            })?
+            .connect_lazy();
+
+        let mut stopwatch = tokio::time::interval(Duration::from_secs(1));
+        let mut aggregate = HashMap::<String, MetricDataPoint>::with_capacity(GCP_MAX_BATCH_SIZE);
+        let mut time_tracker = StartTimeTracker::new();
+        let mut last_time = Instant::now();
+        let mut buffer = Vec::<MetricDataPoint>::new();
+        let mut client =
+            MetricServiceClient::<Channel>::with_interceptor(channel, interceptor(auth));
+
+        loop {
+            if aggregate.len() == GCP_MAX_BATCH_SIZE {
+                while last_time.elapsed() < GCP_SMALLEST_CREATE_TIMESERIES_CALL_FREQUENCY {
+                    tokio::time::sleep(Duration::from_millis(500)).await;
+                }
+            } else {
+                while let Some(data_point) = buffer.pop() {
+                    aggregate
+                        .entry(format!("{}/{}", data_point.namespace, data_point.name))
+                        .and_modify(|cur| cur.merge(&data_point))
+                        .or_insert(data_point);
+
+                    if aggregate.len() == GCP_MAX_BATCH_SIZE {
+                        break;
+                    }
+                }
+            }
 
-    async fn build_request(
-        &self,
-        mut metrics: Self::Output,
-    ) -> crate::Result<hyper::Request<Bytes>> {
-        let metric = metrics.pop().expect("only one metric");
-        let (series, data, _metadata) = metric.into_parts();
-        let namespace = series
-            .name
-            .namespace
-            .unwrap_or_else(|| self.config.default_namespace.clone());
-        let metric_type = format!(
-            "custom.googleapis.com/{}/metrics/{}",
-            namespace, series.name.name
-        );
-
-        let end_time = data.time.timestamp.unwrap_or_else(chrono::Utc::now);
-
-        let (point_value, interval, metric_kind) = match &data.value {
-            MetricValue::Counter { value } => {
-                let interval = gcp::GcpInterval {
-                    start_time: Some(self.started),
-                    end_time,
+            if !aggregate.is_empty()
+                && last_time.elapsed() >= GCP_SMALLEST_CREATE_TIMESERIES_CALL_FREQUENCY
```

Comment:
Could we have a comment here explaining why we need to wait here?

Reply:
Done